src/d3a/models/strategy/area_agents/two_sided_pay_as_clear_engine.py

```diff
@@ -127,63 +127,115 @@ def _match_offers_bids(self):
             self.markets.source.state.clearing[time] = (clearing_rate, clearing_energy)
 
         cumulative_traded_bids = 0
+        accepted_bids = []
         for bid in self.sorted_bids:
-            already_tracked = self.owner.name == bid.buyer
+            original_bid_rate = bid.original_bid_price / bid.energy
             if cumulative_traded_bids >= clearing_energy:
                 break
-            elif (bid.price/bid.energy) >= clearing_rate and \
+            elif (bid.price / bid.energy) >= clearing_rate and \
                     (clearing_energy - cumulative_traded_bids) >= bid.energy:
                 cumulative_traded_bids += bid.energy
-                self.markets.source.accept_bid(
+                trade = self.markets.source.accept_bid(
                     bid=bid,
                     energy=bid.energy,
                     seller=self.owner.name,
-                    already_tracked=already_tracked,
-                    trade_rate=clearing_rate
+                    already_tracked=True,
+                    trade_rate=clearing_rate,
+                    original_trade_rate=original_bid_rate
                 )
-            elif (bid.price/bid.energy) >= clearing_rate and \
+            elif (bid.price / bid.energy) >= clearing_rate and \
                     (0 < (clearing_energy - cumulative_traded_bids) <= bid.energy):
 
-                self.markets.source.accept_bid(
+                trade = self.markets.source.accept_bid(
                     bid=bid,
                     energy=(clearing_energy - cumulative_traded_bids),
                     seller=self.owner.name,
-                    already_tracked=already_tracked,
-                    trade_rate=clearing_rate
+                    already_tracked=True,
+                    trade_rate=clearing_rate,
+                    original_trade_rate=original_bid_rate
                 )
                 cumulative_traded_bids += (clearing_energy - cumulative_traded_bids)
+            accepted_bids.append(trade)
             self._delete_forwarded_bid_entries(bid)
 
         cumulative_traded_offers = 0
         for offer in self.sorted_offers:
-            already_tracked = self.owner.name == offer.seller
             if cumulative_traded_offers >= clearing_energy:
                 break
-            elif (math.floor(offer.price/offer.energy)) <= clearing_rate and \
+            elif (math.floor(offer.price / offer.energy)) <= clearing_rate and \
                     (clearing_energy - cumulative_traded_offers) >= offer.energy:
+                # TODO: Used the clearing_rate as the original_trade_rate for the offers, because
+                # currently an aggregated market is used. If/once a peer-to-peer market is
+                # implemented, we should use the original bid rate for calculating the fees
+                # on the source offers, similar to the two sided pay as bid market.
+
+                # energy == None means to use the bid energy instead of this
```

Comment:
I propose to replace `this` by `offer.energy`

Reply:
Agreed, fixing